### PR TITLE
deploying --only hosting also deploys tagged functions

### PR DESCRIPTION
**Pull request by inlined**
DO NOT MERGE UNTIL API COUNCIL GIVES THE LGTM IN THE DISCUSSION THREAD

Surgical implementation to add any functions specifically tagged to the deploy targets even when --only hosting is specified.

Tested manually in addition to using unit tests

### CODE REVIEW

File: src/deploy/hosting/prepare.ts [hunk 1]
```diff
@@ -23,6 +21,41 @@ export async function prepare(context: Context, options: HostingOptions & Option
...
+  options: HostingOptions & Options
+): Promise<boolean> {
+  // This must be called before modifying hosting config because we turn it from
+  // a scalar to an array now
+  handlePublicDirectoryFlag(options);
+
+  let addedFunctions = false;
+  for (const c of config.hostingConfig(options)) {
+    for (const r of c.rewrites || []) {
+      if (!("function" in r) || typeof r.function !== "object" || !r.function.pinTag) {
```
Comment (jamesdaniels) [inline src/deploy/hosting/prepare.ts:40]: add same logic for r.run.pinTag too

File: src/deploy/index.ts [hunk 2]
```diff
@@ -77,6 +78,16 @@ export const deploy = async function (
     }
   }
 
+  if (
```
Comment (jamesdaniels) [inline src/deploy/index.ts:81]: we need to remove the functions unshift in frameworks/index#prepareFrameworks for this logic to work

File: src/deploy/hosting/prepare.ts (hunk 1 above)
Comment (inlined) [reply src/deploy/hosting/prepare.ts:40]: Don't need it because we don't support deploying a Run service.

File: src/deploy/index.ts (hunk 2 above)
Comment (inlined) [reply src/deploy/index.ts:81]: Oh, really good point. I think we should just remove the guard against `!targetName.includes("functions")` to simplify things.

File: src/deploy/hosting/prepare.ts (hunk 1 above)
Comment (jamesdaniels) [reply src/deploy/hosting/prepare.ts:40]: oh fair point

File: src/deploy/hosting/prepare.ts [hunk 3]
```diff
@@ -23,6 +21,41 @@ export async function prepare(context: Context, options: HostingOptions & Option
...
+        continue;
+      }
+
+      const existing = await backend.existingBackend(context);
+      const endpoint =
+        existing.endpoints[r.function.region || "us-central1"][r.function.functionId];
+      options.only = ensureTargeted(options.only, endpoint.codebase || "default", endpoint.id);
```
Comment (jamesdaniels) [inline src/deploy/hosting/prepare.ts:47]: only might be undefined, can handle either here or default in ensureTargeted

File: src/deploy/index.ts (hunk 2 above)
Comment (jamesdaniels) [reply src/deploy/index.ts:81]: removing the guard would break `--only hosting,functions` as it would then append to only

Comment (inlined) [reply src/deploy/index.ts:81]: Actually, we need to remove the guard. 

`--only hosting,functions` will still be `--only hosting,functions` after ensureTargeted, but `--only hosting,functions:someOtherFunction` may need to be `--only hosting,functions:ssr,functions:someOtherFunction`

**Review by bkendall — APPROVED** [review]

**Review by jamesdaniels — APPROVED** [review]

**Pull request merged**